File: src/main/scala/2.add-two-numbers.scala
/**
  * Definition for singly-linked list.
  * class ListNode(var _x: Int = 0) {
  *   var next: ListNode = null
  *   var x: Int = _x
  * }
  */
class ListNode(var _x: Int = 0) {
  var next: ListNode = _
  var x: Int = _x
}
object Q2 {
  def addTwoNumbers(l1: ListNode, l2: ListNode): ListNode = {
    var acc: ListNode = null
    var init: ListNode = null
    var carry = 0
    @scala.annotation.tailrec
    def loop(l1: ListNode, l2: ListNode): ListNode = {
      if (l1 == null && l2 == null) {
        return init
      }
      val ls1 = if (l1 == null) new ListNode(0) else l1
      val ls2 = if (l2 == null) new ListNode(0) else l2
      val sum = carry + ls1._x + ls2._x
      val curr = new ListNode(sum % 10)
      carry = sum / 10

      if (acc == null) acc = curr
      else {
        acc.next = curr
        acc = acc.next
      }
      if (init == null) init = curr
      loop(ls1.next, ls2.next)
    }
    val ret = loop(l1, l2)
    // 如果还有进位, 则也要把进位加上
    if (carry > 0) {
      acc.next = new ListNode(carry)
    }
    ret
  }
}

File: src/main/java/2.add-two-numbers.java
import common.java.ListNode;

/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode(int x) { val = x; } }
 */
class Q2Java {
  public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
    ListNode head = new ListNode(-1);
    ListNode ret = head;
    boolean carry = false;
    while (l1 != null || l2 != null) {
      int val1 = l1 == null ? 0 : l1.val;
      int val2 = l2 == null ? 0 : l2.val;
      int sum = val1 + val2;
      if (carry) {
        sum += 1;
        // 先重置
        carry = false;
      }
      if (sum >= 10) {
        sum = sum % 10;
        // 又有进位则再进位
        carry = true;
      }
      head = addNode(head, sum);
      if (l1 != null)
        l1 = l1.next;
      if (l2 != null)
        l2 = l2.next;
    }
    // 循环出来之后carry还存在, 需要再进一位
    if (carry) {
      head = addNode(head, 1);
    }
    return ret.next;
  }

  private ListNode addNode(ListNode target, int number) {
    ListNode newNode = new ListNode(number);
    target.next = newNode;
    return newNode;
  }
}
